### ISSUE TITLE
[WIP] Clean up code smells reported by CodeClimate.

### ISSUE THREAD

**Issue by jdickey** [jdickey/new_poc#162]
See the reports for [`Newpoc::Repository`](https://codeclimate.com/github/jdickey/new_poc/Newpoc::Repository) and [`Newpoc::Repository::Internal`](https://codeclimate.com/github/jdickey/new_poc/Newpoc::Repository::Internal).

Score dropped from 3.98 to 3.66; not good. However, it's not clear from the linked report pages if CodeClimate is complaining about the complexity of the *code* or the complexity of the *specs*.

**Comment by jdickey** [comment]
One of the CodeClimate quirks that makes more sense after thinking about it for a while is in how it reports code duplication: with the error report (and score hit) repeated once for each *repetition*, not for each *item* that's duplicated. If you repeat the same chunk of code 20 times, it *should* bring your score down proportionately more than if you have a chunk repeated twice. Oh, well.

**Comment by jdickey** [comment]
It's really hard to see what (the free version of) CodeClimate is complaining about sometimes, for example with its `F` grade for the [`Newpoc::Entity::SupportClasses` module](https://codeclimate.com/github/jdickey/new_poc/Newpoc::Entity::Post::SupportClasses), which in fact contains four classes [`BylineBuilder`](https://github.com/jdickey/new_poc/blob/52915bd/lib/newpoc-entity-post/lib/newpoc/entity/post/byline_builder.rb), [`ImageBodyBuilder`](https://github.com/jdickey/new_poc/blob/52915bdc4913f6d527b5272f8a21631bee47e9b3/lib/newpoc-entity-post/lib/newpoc/entity/post/image_body_builder.rb), [`TextBodyBuilder`](https://github.com/jdickey/new_poc/blob/52915bdc4913f6d527b5272f8a21631bee47e9b3/lib/newpoc-entity-post/lib/newpoc/entity/post/text_body_builder.rb), and [`TimestampBuilder`](https://github.com/jdickey/new_poc/blob/52915bdc4913f6d527b5272f8a21631bee47e9b3/lib/newpoc-entity-post/lib/newpoc/entity/post/timestamp_builder.rb)). Not sure how to proceed further there; hence, marking this issue as a WIP.

**Comment by jdickey** [comment]
As of Commit 1f7f673, we're back up to 3.98 GPA, which rounds on the README badge to 4.0.  CodeClimate lists (only?) two [duplication issues](https://codeclimate.com/github/jdickey/new_poc/issues/categories/duplication) (that should be addressed) and one [complex method](https://codeclimate.com/github/jdickey/new_poc/issues/categories/complexity) (which I plan on leaving as is, at least for now).

In [this comment](https://github.com/jdickey/new_poc/issues/54#issuecomment-77142902) on Issue #54, I said that I'd defer "[looking] at factoring out common behaviour between Post and User actions" until this issue. Here we are.

In [this comment](https://github.com/jdickey/new_poc/pull/210#issuecomment-76625125) on PR #210, I said I should factor out common code between [`PostsController::Action::Create::Internals::PostDataFilter`](https://github.com/jdickey/new_poc/blob/1f7f673/app/controllers/posts_controller/action/create/internals/post_data_filter.rb) and [`UsersController::Action::Update::Internals::UserDataFilter`](https://github.com/jdickey/new_poc/blob/1f7f673/app/controllers/users_controller/action/update/internals/user_data_filter.rb) as part of this issue. Here we are.

Issue #169 has *significant* overlap with this issue.

#### Proposed ordering of the above

1. Eliminate the duplication called out by CodeClimate; granted, that'll only bring us up to 3.9*9*, but still;
1. Factor out common code as called out in the Issue #54 comment;
1. Spend a day or two looking at the specs and code to find *sensible* DRYing up/opportunities for custom matchers and/or support methods.

Then we can close this issue and go on to other issues &mdash; likely #211, #169 and #203. *That* would close all currently-open issues not labelled either `fixlater` or `wontfix` *except for* #54 — which is pending on #232, really.

Thoughts?

**Comment by jdickey** [comment]
Well, PR #237 din't make CodeClimate any happier; it now [complains](https://codeclimate.com/github/jdickey/new_poc/issues/categories/complexity) that our newly-added [`def_action` class method](https://codeclimate.com/github/jdickey/new_poc/ApplicationController) is "complex".

CodeClimate also doesn't like branch names that have been previously used; PR #237 merged an `issue-162` branch, which had previously been used (and pushed to GitHub, and seen by CodeClimate). To have CC evaluate a branch, it really needs to have a name that hasn't been used before. That's suboptimal, IMO.

**Comment by mitpaladin** [comment]
Indeed, rather picky, though to make it happy we could use the issue code + short descriptor.  Just to keep CC happy.

**Comment by jdickey** [comment]
Now we're back to [complex code warnings](https://codeclimate.com/github/jdickey/new_poc/issues/categories/complexity) relating to the calls to `def_method` in `PostsController` and `UsersController`.

Hang on... a bit of digging shows that [RuboCop's default threshold](https://github.com/bbatsov/rubocop/blob/master/.rubocop_todo.yml#L9-L10) for the ABC metric is 38. We've just established that CodeClimate's threshold is *30*. We can't tell CodeClimate to ease up (though I've just sent off a support email suggesting that); we *can* set RuboCop's config to match, so we don't have to commit and push just to see if CodeClimate is going to complain about complexity.

:tired_face:

**Comment by jdickey** [comment]
Right, then; none of the changes I've made in the last 4 hours or so (up to Commit 8fabe338) have helped [the CodeClimate score](https://codeclimate.com/github/jdickey/new_poc/PostsController?from_sha=1f7f6738&to_sha=8fabe338) for [`posts_controller.rb`](https://github.com/jdickey/new_poc/blob/8fabe338/app/controllers/posts_controller.rb); it's still complaining of "complex code...outside of methods". It *apparently* doesn't like the *calls to* the controller class' shiny new [`def_method`](https://github.com/jdickey/new_poc/blob/8fabe338/app/controllers/application_controller.rb#L54-L62) class method, which (as a reminder) turns

```ruby
  def update
    Action::Update.new(current_user: current_user, slug: params[:id],
                       repository: PostRepository.new,
                       post_data: params[:post_data])
      .subscribe(self, prefix: :on_update)
      .execute
  end
```

into

```ruby
  def_action(:update) do
    {
      current_user: current_user, slug: params[:id],
      post_data: params[:post_data], repository: PostRepository.new
    }
  end
```

I'd also point out that CodeClimate's claimed ABC complexity score for that class (37) is *significantly* higher than the limit we've set in our [RuboCop config](https://github.com/jdickey/new_poc/blob/8fabe338495456245436027d820b074cadd8e638/.rubocop.yml#L4) (20). Somebody doesn't think those numbers mean what somebody else thinks they mean, and I'm done trying to mediate.

I'm going to bail out and declare that 3.95 is hereby an acceptable CodeClimate overall score; it's not worth throwing extra time at it beyond what's there now.

**Comment by jdickey** [comment]
Moral of the story, if it isn't obvious by now: [RuboCop](https://github.com/bbatsov/rubocop) and [Flog](https://github.com/seattlerb/flog) have very different ideas about [ABC metrics](http://www.c2.com/cgi/wiki?AbcMetric). CodeClimate's "complexity" sniff tests are based on Flog, which

* treats `require`, *and the accumulation of the original file with all `require`d files*, as significantly complex; and
* is not adequately detailed in the CodeClimate reports, which merely list "complexity in *filename* outside any methods".

**tl;dr** It's apparently impossible to get a CodeClimate score higher than the 3.95-3.98 range for any project with real-world separation of cooperating classes; its flawed treatment of `require` prevents it.
